** Install and use

*1. Download and install workflow*

#+begin_src sh
git clone https://github.com/thackl/detectEVE
cd detectEVE

# install with conda - I prefer mamba via https://github.com/conda-forge/miniforge
mamba create -n detectEVE
mamba activate detectEVE
mamba env update --file workflow/envs/env.yaml
#+end_src

*2. Setup databases*

#+begin_src sh
# download and cluster rvdb
# download uniref50 for retroblast
./detectEVE --setup-databases
#+end_src

*3. Analyse genomes*
#+begin_src sh
# from files
./detectEVE *.fasta

# from NCBI SRA WGS accessions
./detectEVE -a ACC [ACC ...]
#+end_src

- for *accessions*: The genome fasta files of all SRA WGS accessions will
  automatically be downloaded from NCBI to =genomes/<accession>.fna= (see
  https://www.ncbi.nlm.nih.gov/Traces/wgs/, for example, QMGA01 for the
  bottlenose dolphin).

*NOTE:* If you encounter an error related to
/tidyverse/stringi/libicui18n.so.58/, try reinstalling =stringi= locally

#+begin_src sh
mamba remove r-stringi r-tidyverse
R -e 'install.packages("stringi")'
mamba install r-tidyverse
#+end_src

To restart the workflow from where it failed, just run the same command again.

** Output
The pipeline produces the following final files in =results/=:
- =<genome_id>-validatEVEs.tsv= - best hit of the EVE with evidence and confidence
  annotation (exogenous virus: high, eve-ish description & hypothetical protein:
  low)
- =<genome_id>-validatEVEs.fna= - validatEVEs nucleotide sequences
- =<genome_id>-validatEVEs.pdf= - graphical overview of hit distribution for validatEVEs
[[file:workflow/detectEVE-output-example.png]]
** Workflow
[[file:workflow/detectEVE-workflow.png]]

