** Usage

To identify EVEs take these 3 steps:

1. setup the databases (see [[Install & Setup]])

2. Provide genomes to scan. You can do so in two ways, via
  - files :: place your genomes as fasta files with =.fna= extensions into the
    =genomes/= folder
  - accessions :: add SRA WGS accessions to the first column of
    =genomes.tsv=. The genome fasta files will automatically be downloaded from
    NCBI to =genomes/<accession>.fna= (see
    https://www.ncbi.nlm.nih.gov/Traces/wgs/, for example, QMGA01 for the
    bottlenose dolphin).

3. Run the pipeline
#+begin_src sh
snakemake --cores 30
#+end_src

The pipeline produces the following final files in =results/=:
- =<genome_id>-validatEVEs.tsv= :: best hit of the EVE with evidence and confidence
  annotation (exogenous virus: high, eve-ish description & hypothetical protein:
  low)
- =<genome_id>-validatEVEs.pdf= :: graphical overview of hit distribution for validatEVEs
- =<genome_id>-validatEVEs.fna= :: validatEVEs nucleotide sequences

** Install & Setup

#+begin_src sh
workflow=omx-discover-eves
git clone https://github.com/thackl/$workflow
cd $workflow
mamba create -n $workflow;          
mamba env update -n $workflow --file workflow/envs/env.yaml
mamba activate $workflow

# download uniref50 for retroblast
snakemake --cores 30 resources/uniref50.dmnd

# add virus ref db
resources/VDB.dmnd
#+end_src

** Changes (compared to legacy)
- =diamond blastx= instead of =tblastn=

- =diamond= is used in frameshift mode to can extend hits across frameshifts;
  dropped merging of abutting =tblastn= hits from orig pipeline

- need to chop long contigs for diamond or else I lose hits (50k + 10k overlap
  windows). Creates possibility of redundant hits right at window edge (could be
  filtered?)

- uniref50 instead of NR for retroblast (faster, less redundant)

- retroblast: --top 20 (top hits within 20% top score, rather than strict -k 10 top hits)

- min_length: 83 amino-acids (250 bp in tblastn)

** Questions & Notes
- lots of putatEVEs are low-complexity repeats - meaningful?

- currently I'm not filtering out retroviral hits - but we probably should?

- are validatEVEs just subset of putatEVEs or somehow extracted based on
  retroblast results?
